engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java

```diff
@@ -0,0 +1,163 @@
+package org.enso.interpreter.runtime.data;
+
+import com.oracle.truffle.api.interop.ArityException;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.interop.UnsupportedTypeException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.dsl.Builtin;
+import org.enso.interpreter.node.expression.builtin.error.PolyglotError;
+import org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.error.DataflowError;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+@Builtin(pkg = "immutable", stdlibName = "Standard.Base.Data.Vector.Vector")
+public final class Vector implements TruffleObject {
+  private final Object storage;
+
+  private Vector(Object storage) {
+    assert InteropLibrary.getUncached().hasArrayElements(storage)
+        : "Vector needs array-like delegate, but got: " + storage;
+    this.storage = storage;
+  }
+
+  @Builtin.Method(
+      name = "new_builtin",
+      description = "Creates new Vector with given length and provided elements.")
+  @Builtin.Specialize
+  public static Object newFromFunction(long length, Function fun, InteropLibrary interop) {
+    Object[] target = new Object[(int) length];
+    for (int i = 0; i < length; i++) {
+      try {
+        final Object value = interop.execute(fun, (long) i);
+        if (value instanceof DataflowError) {
+          return value;
+        }
+        target[i] = value;
+      } catch (ArityException | UnsupportedTypeException | UnsupportedMessageException e) {
+        throw raise(RuntimeException.class, e);
+      }
+    }
+    return new Vector(new Array(target));
+  }
+
+  @SuppressWarnings("unchecked")
+  private static <E extends Exception> E raise(Class<E> clazz, Throwable t) throws E {
+    throw (E) t;
+  }
+
+  /**
+   * Marks the object as array-like for Polyglot APIs.
+   *
+   * @return {@code true}
+   */
+  @ExportMessage
+  public boolean hasArrayElements() {
+    return true;
+  }
+
+  /**
+   * Exposes the size of this collection through the polyglot API.
+   *
+   * @return the size of this array
+   */
+  @ExportMessage
+  public long getArraySize() throws UnsupportedMessageException {
+    InteropLibrary interop = InteropLibrary.getUncached();
+    return interop.getArraySize(storage);
+  }
```

Comment:
```suggestion
  public long getArraySize(@CachedLibrary("storage") InteropLibrary interop) throws UnsupportedMessageException {
    return interop.getArraySize(storage);
  }
```

Reply:
Done in [f0338c5](https://github.com/enso-org/enso/pull/3667/commits/f0338c5e1a2a2621db4cf4cbc802a7b0d746cca5)